File: C#/46. Permutations.cs
public class Solution
{
    private List<IList<int>> DFS(List<int> list)
    {
        List<IList<int>> ans = new List<IList<int>>();
        if (list.Count == 1)
        {
            ans.Add(new List<int>() { list[0] });
            return ans;
        }
        int n = list[0];
        list.RemoveAt(0);
        List<IList<int>> allPer = DFS(list);
        foreach (IList<int> p in allPer)
        {
            for (int i = 0; i <= p.Count; ++i)
            {
                List<int> newP = new List<int>(p);
                newP.Insert(i, n);
                ans.Add(newP);
            }
        }
        return ans;
    }
    public IList<IList<int>> Permute(int[] nums)
    {
        return DFS(nums.ToList());
    }
}

File: python/46. Permutations.py
# -*- coding: utf-8 -*-

"""
@Author: xiezizhe
@Date: 14/6/2020 下午6:48
"""
from typing import List


class Solution:

    def dfs(self, nums: List[int], idx: int) -> List[List[int]]:
        if len(nums) <= idx + 1:
            return [[nums[idx]]]

        f = nums[idx]
        res = []
        for p in self.dfs(nums, idx + 1):
            for i in range(1, len(p)):
                res.append(p[:i] + [f] + p[i:])
            res.append([f] + p)
            res.append(p + [f])

        return res

    def permute(self, nums: List[int]) -> List[List[int]]:
        if nums is None or len(nums) == 0:
            return [[]]

        if len(nums) == 1:
            return [nums]

        return self.dfs(nums, 0)


if __name__ == "__main__":
    s = Solution()
    print(s.permute([1]))
